### ISSUE TITLE
How should destination distance restrict which mission is chosen ?

### ISSUE THREAD

**Issue by mokun** [mars-sim/mars-sim#904]
**Is your feature request related to a problem? Please describe.**
- Prior to starting a mission, a few criteria should be sketched out right away, such as the number of destinations and projected distance and projected cargo weight. This is called **Mission Planning**. 

**Describe the solution you desire to see**
- Ideally speaking, a vehicle's base range would restrict how far it can travel. But the fuel consumption can be very tricky since high resolution detailed survey of terrain elevation on Mars is not available.

**Describe the rationale**
- A vehicle should not be chosen, only to find out later that it doesn't have the range to complete the journey. Destination distance should restrict a vehicle from being chosen in the first place.

**Codes**
- In some meta mission class (but not all meta mission), it uses :
```java 
// Get available rover.
Rover rover = RoverMission.getVehicleWithGreatestRange(settlement, false);
```
- A mission should choose the next best rover, NOT a rover with greatest range. This would prevent another mission that needs this vehicle to be cancelled.

- Also, all rover missions should always first decide on the destination and its distance as a criteria for the mission.

**Specifications  (please complete the following information):**
 - pre 3.6.0 build 8132

**Comment by mokun** [comment]
Will undo a small part of https://github.com/mars-sim/mars-sim/commit/7b92056a53ff314d5b310c3e30d53cbbca9453a8 that prevent any missions from starting. 

The reason was that `distanceProposed` at the moment of reserving a vehicle is always zero.

**Comment by bevans2000** [comment]
> Will undo a small part of [7b92056](https://github.com/mars-sim/mars-sim/commit/7b92056a53ff314d5b310c3e30d53cbbca9453a8) that prevent any missions from starting.
> 
> The reason was that `distanceProposed` at the moment of reserving a vehicle is always zero. So it won't work.

Yes, the vehicle is chosen before the  mission destination. The mission locations are based on the calculated range of the vehicle. So I noticed that recently ranges have become very large maybe due to cha ge for Vehiclecontroller.

**Comment by bevans2000** [comment]
The capping of distance has to be in the constructor of the specific missions. Either

- percentage of vehicle range
- maximum mission duration, this is harder
- fixed value for each mission

**Comment by mokun** [comment]
Just committed [this](https://github.com/mars-sim/mars-sim/commit/ac23b5e4a211e5aac25dcaf7fe26886a504532d8) to remove that check.

**Comment by mokun** [comment]
> So I noticed that recently ranges have become very large maybe due to cha ge for Vehiclecontroller.

Yes I've toned down the base range. The calculation was a bit off. It's more reasonable now.

In near future, I'll look up theories on vehicle dynamic and add in a few more params for predicting the drive range better. For now, there's still quite a bit of hacks with handcrafted values.

**Comment by mokun** [comment]
Actually, it's also possible a mission planning to do the parallel of both looking for just available vehicle to see what range it can provide before limiting the visiting locations.

Of course, if there are multiple vehicles available, then it's best to first decide where to go, then select the right vehicle with enough range.

So both approaches are valid.

**Comment by bevans2000** [comment]
In #890 I have followed the vehicle first and then mission route second. I think that makes sense but we need to cap the distance when searching for destinations

**Comment by bevans2000** [comment]
As part of #890 I am splitting up the various logic blocks to separate class/methods which will give us more flexibility.
I want introduce a generic VehicleScorer that can assess a vehicle against a Mission depending on the requirements needed. This will be essential to allow custom Mission to be defined.

**Comment by mokun** [comment]
We could use the RACI model that @Urwumpe was alluring to potentially crafting better mission. See #226. It was still open.

**Comment by Urwumpe** [comment]
That one is mostly architectual and only prevents, that too many Settlers do the same, worst of all, doing decisions.

Here, I would rather go having criteria, that could compare vehicles to other vehicles or vehicles to missions. And allow to add criteria to a decision/Rover selection process as needed by the mission.

Can we allow alternative mission plans to a mission in the planning phase? Like a transport mission using a convoy of multiple rovers or fit all into a single big Rover? This could result in more variance of missions with fewer dedicated mission classes and allow fitting a mission better to the capabilities of a smaller or larger settlement. Just the special components of how a mission is planned would need to be hard-coded, like what constitutes a fully-planned mineral exploration mission.

**Comment by bevans2000** [comment]
I think where we are trying to go with #890 of making Missions be a sequence of Steps should give us more flexibility. The Steps become individual building blocks.
So for multiple vehicles we could create a Mission of Missions; each individual Mission would be control a single Vehicle. But the top level Mission would be about doing a large scale trade with a settlement. Same for Construction as well with the sub-mission being the individual parts, foundation, frame etc. The top level Mission would be the Plan.

**Comment by mokun** [comment]
Back to the original question regarding the distance, we've since made some progress in making the proposed mission to calculating shorter distance.

However, from time to time, long distance still pops up and got approved.   

In this Alpha Base settlement called Tiberius, in the first 2 sols, they proposed over 12 missions. 4 of them have been approved as follows: 

![image](https://github.com/mars-sim/mars-sim/assets/1168584/78d3644e-0d2d-4f71-8eac-396aed056cd4)

The last one (a regolith mission with 34 km only) did not get approved. Then on sol 3, a new one (a mining mission with 664 km) got approved. See below: 

![image](https://github.com/mars-sim/mars-sim/assets/1168584/0dd72343-8cd2-4f3e-abc8-323993ac5967)

To fix this, I'm thinking of differentiating the opinion score between divisions. For instance, 
1. The safety specialists should be the one who would caution against long distance since the longer the distance, the higher the risk factors are. 
2. The logistic specialists would score on how likely the mission plans would adhere to the EVA time and travel time.
3. The resource specialists would focus on scoring the food/water/fuel consumable
4. The science specialists would comment on the scientific returns on investment

Do you guys like that idea ?

**Comment by bevans2000** [comment]
So are you saying that a Mission has to be score by multiple Persons, each with a different Role?

**Comment by mokun** [comment]
> a Mission has to be score by multiple Persons, each with a different Role?

Yes. This is how it's done traditionally. It needs to be reviewed by multiple people. Each decision will have a weight of 10 to 20%. If the total score is larger than the last passing score, then the mission is approved.

But now, we want to add a new twist to it. 

Since each reviewer has a role, we may as well take advantage of that by having them doing their job based upon their role perspective.

**Comment by mokun** [comment]
@bevans2000 @Urwumpe 

During the course of reworking the Exploration mission, have you by any chance already reduced the site distance, especially at the beginning ? 

As times go on, it makes sense to unlock progressively farther distance, given the accumulated mission experiences.

Having said that, how would you quantify the mission experiences received by the settlers in terms of numbers ? 

Where would you store these experiences ? 

We don't have a Mission skill type yet.

**Comment by bevans2000** [comment]
Generally the range of the vehicle constraints the distance. There is experience recorded against the Settlement for Missions. I don't mind tail9ring the distance but it must be based on experience not the mission sol.

**Comment by mokun** [comment]
So how would you compute the mission experience ? 

We will have to figure out how to store this number on each mission member also.

**Comment by bevans2000** [comment]
At the end of a mission allembers get an experience uplift. So there is something recorded.
Let's tag this for 4.2 and pick it up

**Comment by mokun** [comment]
> an experience uplift

Prior to the mission, there's a mission qualification score when executing ReviewMissionPlan.

In AbstractMetaMission,
```java
	/**
	 * Gets the mission qualification value for the member. Member is qualified in
	 * joining the mission if the value is larger than 0. The larger the
	 * qualification value, the more likely the member will be picked for the
	 * mission.
	 *
	 * @param member the member to check.
	 * @return mission qualification value.
	 */
	@Override
	public double getWorkerSuitability(Worker member) {

		double result;

		if (member instanceof Person person) {
			result = Math.max(5,  person.getMissionExperience(type));

			// Get base result for job modifier.
			JobType job = person.getMind().getJobType();
			double jobModifier = 2D;
			if (preferredWorkerJob.isEmpty() || preferredWorkerJob.contains(job)) {
				jobModifier = 3D;
			}

			result = result * jobModifier;
		}
		else {
			Robot robot = (Robot) member;

			// Get base result for job modifier.
			result = (preferredRobots.contains(robot.getRobotType())) ? 30.0 : 0;
		}

		return result;
	}

```

So there is indeed already a mission experience score for workers that record experiences on each type of mission.

In Person,
```java
	/** The person's mission experiences. */
	private EnumMap<MissionType, Integer> missionExperiences;

...
	/**
	 * Adds the mission experience score.
	 *
	 * @param missionType
	 * @param score
	 */
	public void addMissionExperience(MissionType missionType, int score) {
		Integer total = missionExperiences.getOrDefault(missionType, 0);
		total += score;
		missionExperiences.put(missionType, total);
	}

	/**
	 * Gets the mission experiences map.
	 *
	 * @return a map of mission experiences
	 */
	public int getMissionExperience(MissionType missionType) {
		Integer previous = missionExperiences.get(missionType);
		if (previous != null) {
			return previous;
		}
		return 0;
	}
```

In AbstractMission,
```java
	/**
	 * Computes the mission experience score.
	 *
	 * @param reason
	 */
	private void addMissionScore() {
		for (Worker member : members) {
			if (member instanceof Person person && !person.isDeclaredDead()) {
				if (person.getPhysicalCondition().hasSeriousMedicalProblems()) {
					// Note : there is a minor penalty for those who are sick
					// and thus unable to fully function during the mission
					person.addMissionExperience(missionType, 2);
				}
				else if (person.equals(startingMember)) {
					// The mission lead receive extra bonus
					person.addMissionExperience(missionType, 6);
					// Add a leadership point to the mission lead
					person.getNaturalAttributeManager().adjustAttribute(NaturalAttributeType.LEADERSHIP, 1);
				}
				else
					person.addMissionExperience(missionType, 3);
			}
		
		}
	}
```

**Comment by mokun** [comment]
> Generally the range of the vehicle constraints the distance

Then it's not enough. We could use the mission experience to moderate the distance to be travelled.

The higher the mission experience, the farther one can travel.
